core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/signer/BaseAwsV4HttpSigner.java

```diff
@@ -0,0 +1,359 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.signer;
+
+import static software.amazon.awssdk.http.auth.aws.util.HttpChecksumUtils.createSdkChecksumFromRequest;
+import static software.amazon.awssdk.http.auth.aws.util.SignerConstant.AWS4_SIGNING_ALGORITHM;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.addHostHeader;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.deriveSigningKey;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.hashCanonicalRequest;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.auth.aws.AwsV4HttpSigner;
+import software.amazon.awssdk.http.auth.aws.checksum.ContentChecksum;
+import software.amazon.awssdk.http.auth.aws.checksum.SdkChecksum;
+import software.amazon.awssdk.http.auth.aws.internal.util.DigestComputingSubscriber;
+import software.amazon.awssdk.http.auth.aws.util.CredentialUtils;
+import software.amazon.awssdk.http.auth.aws.util.HttpChecksumUtils;
+import software.amazon.awssdk.http.auth.aws.util.SignerConstant;
+import software.amazon.awssdk.http.auth.aws.util.SignerUtils;
+import software.amazon.awssdk.http.auth.spi.AsyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.AsyncSignedRequest;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignedRequest;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.utils.BinaryUtils;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.StringUtils;
+
+/**
+ * An internal extension of {@link AwsV4HttpSigner} that enables composable implementations of aws-signers that use
+ * a set of properties, which may extend {@link AwsV4Properties}, in order to sign requests.
+ * <p>
+ * The process for signing requests to AWS services is documented
+ * <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_aws-signing.html">here</a>.
+ */
+@SdkProtectedApi
+public interface BaseAwsV4HttpSigner<T extends AwsV4Properties> extends AwsV4HttpSigner {
```

Comment:
This is kind of a large protected API. I'll need to take some time to understand if there's a better abstraction once we've gotten checksums out of here.